cpp.nim

```diff
@@ -254,48 +258,64 @@ proc parseIfndef(p: var Parser; sectionParser: SectionParser): PNode =
   result = ast.emptyNode
   getTok(p) # skip #ifndef
   expectIdent(p)
-  if p.tok.s == c2nimSymbol:
-    skipLine(p)
-    case skipUntilElifElseEndif(p)
-    of emElif:
-      result = newNodeP(nkWhenStmt, p)
-      addSon(result, newNodeP(nkElifBranch, p))
-      getTok(p)
-      addSon(result.sons[0], expression(p))
-      eatNewLine(p, nil)
-      parseIfDirAux(p, result, sectionParser)
-    of emElse:
-      skipLine(p)
-      result = parseStmtList(p, sectionParser)
-      eatEndif(p)
-    of emEndif: skipLine(p)
+  if p.options.skipped.contains(p.tok.s):
+    skipUntilEndif(p)
+    result = ast.emptyNode
   else:
-    # test if include guard:
-    saveContext(p)
-    if isIncludeGuard(p):
-      closeContext(p)
-      result = parseStmtList(p, sectionParser)
-      eatEndif(p)
+    if p.tok.s == c2nimSymbol:
+      skipLine(p)
+      case skipUntilElifElseEndif(p)
+      of emElif:
+        result = newNodeP(nkWhenStmt, p)
+        addSon(result, newNodeP(nkElifBranch, p))
+        getTok(p)
+        addSon(result.sons[0], expression(p))
+        eatNewLine(p, nil)
+        parseIfDirAux(p, result, sectionParser)
+      of emElse:
+        skipLine(p)
+        result = parseStmtList(p, sectionParser)
+        eatEndif(p)
+      of emEndif: skipLine(p)
     else:
-      backtrackContext(p)
-      result = newNodeP(nkWhenStmt, p)
-      addSon(result, newNodeP(nkElifBranch, p))
-      var e = newNodeP(nkPrefix, p)
-      addSon(e, newIdentNodeP("not", p))
-      addSon(e, definedExprAux(p))
-      eatNewLine(p, nil)
-      addSon(result.sons[0], e)
-      parseIfDirAux(p, result, sectionParser)
+      # test if include guard:
+      saveContext(p)
+      if isIncludeGuard(p):
+        closeContext(p)
+        result = parseStmtList(p, sectionParser)
+        eatEndif(p)
+      else:
+        backtrackContext(p)
+        result = newNodeP(nkWhenStmt, p)
+        addSon(result, newNodeP(nkElifBranch, p))
+        var e = newNodeP(nkPrefix, p)
+        addSon(e, newIdentNodeP("not", p))
+        addSon(e, definedExprAux(p))
+        eatNewLine(p, nil)
+        addSon(result.sons[0], e)
+        parseIfDirAux(p, result, sectionParser)
+
+proc definedGuard(n: PNode): string =
+  if (not n.sons.isNil) and (n.sons.len == 2):
+    let call = n.sons[0]
+    if call.ident.s == "defined":
+      result = n.sons[1].ident.s
 
 proc parseIfDir(p: var Parser; sectionParser: SectionParser): PNode =
   result = newNodeP(nkWhenStmt, p)
   addSon(result, newNodeP(nkElifBranch, p))
   getTok(p)
-  addSon(result.sons[0], expression(p))
-  eatNewLine(p, nil)
-  parseIfDirAux(p, result, sectionParser)
-  if pfAssumeIfIsTrue in p.options.flags:
-    result = result.sons[0].sons[1]
+  let condition = expression(p)
+  let guard = definedGuard(condition)
+  if (guard == "__cplusplus") or p.options.skipped.contains(guard):
```

Comment:
Ok, but I will need to distinguish between `ifdef` to be skipped (which include `__cplusplus`) and `ifndef` to be skipped (which don't)